From f7d5354f0b8d5cd3c4bc71021b83f84ffe3dbd2e Mon Sep 17 00:00:00 2001
From: Bin Jiang <bin.jiang@windriver.com>
Date: Thu, 19 Mar 2015 03:14:19 +0000
Subject: [PATCH 1/2] MIPS: Octeon: When down a CPU, disable local interrupt
 after the irqs are migrated.

When down a CPU, irq_work will be notified to handle all works. irq_work
should work in hard irq disabled environment, otherwise there are following
calltrace:

Kernel bug detected[#1]:
CPU: 12 PID: 55 Comm: migration/12 Not tainted 3.14.29ltsi-WR7.0.0.0_standard #1
task: 8000000409dc9b00 ti: 8000000409df0000 task.ti: 8000000409df0000
$ 0 : 0000000000000000 ffffffff809b6704 0000000000000001 800000002442d188
$ 4 : ffffffff80f83570 0000000000000008 000000000000000c ffffffff809ce9f8
$ 8 : 800000040a678000 ffffffff809ce978 ffffffff809ce9dc 0000000000000019
$12 : 8000000409df3bc8 0000000000009c00 ffffffffac000000 8000000402650000
$16 : ffffffff80e98188 ffffffff80c124f8 0000000000000000 0000000000000008
$20 : 000000000000000c 0000000000000001 0000000000000001 0000000014009ce1
$24 : 0000000003bf0000 ffffffff801d68e8
$28 : 8000000409df0000 8000000409df3b70 8000000409df3b70 ffffffff8026ddec
Hi : 0000000000001c1c
Lo : 0000000000000065
epc : ffffffff8026de14 __irq_work_run+0x84/0x148
    Not tainted
ra : ffffffff8026ddec __irq_work_run+0x5c/0x148

So after the irqs are migrated, disable the local interrupt.
The delay period is used for pending irqs.

Signed-off-by: Bin Jiang <bin.jiang@windriver.com
---
 arch/mips/cavium-octeon/smp.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/arch/mips/cavium-octeon/smp.c b/arch/mips/cavium-octeon/smp.c
index b8ffcd2..3d96cb5 100644
--- a/arch/mips/cavium-octeon/smp.c
+++ b/arch/mips/cavium-octeon/smp.c
@@ -340,6 +340,8 @@ static int octeon_cpu_disable(void)
 	local_irq_disable();
 	octeon_fixup_irqs();
 	local_irq_enable();
+	mdelay(1);
+	local_irq_disable();
 
 	flush_cache_all();
 	local_flush_tlb_all();
-- 
1.7.5.4

